test/adapter/json/nested_relationships_test.rb

```diff
@@ -0,0 +1,203 @@
+require 'test_helper'
+
+NestedPostSerializer = Class.new(ActiveModel::Serializer) do
+  attributes :id, :title
+
+  has_many :comments, include: :author
+end
+
+NestedCommentBelongsToSerializer = Class.new(ActiveModel::Serializer) do
+  attributes :id, :body
+
+  belongs_to :author, include: [:posts]
+end
+
+NestedAuthorSerializer = Class.new(ActiveModel::Serializer) do
+  attributes :id, :name
+
+  has_many :posts, include: [:comments]
+end
+
+ComplexNestedAuthorSerializer = Class.new(ActiveModel::Serializer) do
+  attributes :id, :name
+
+  # it would normally be silly to have this in production code, cause a post's
+  # author in this case is always going to be your root object
+  has_many :posts, include: [:author, comments: [:author]]
+end
+
+module ActiveModel
+  class Serializer
+    class Adapter
+      class Json
+        class NestedRelationShipsTestTest < Minitest::Test
+          def setup
+            ActionController::Base.cache_store.clear
+            @author = Author.new(id: 1, name: 'Steve K.')
+
+            @post = Post.new(id: 42, title: 'New Post', body: 'Body')
+            @first_comment = Comment.new(id: 1, body: 'ZOMG A COMMENT')
+            @second_comment = Comment.new(id: 2, body: 'ZOMG ANOTHER COMMENT')
+
+            @post.comments = [@first_comment, @second_comment]
+            @post.author = @author
+
+            @first_comment.post = @post
+            @second_comment.post = @post
+
+            @blog = Blog.new(id: 1, name: "My Blog!!")
+            @post.blog = @blog
+            @author.posts = [@post]
+          end
+
+          def test_complex_nested_has_many
+            @first_comment.author = @author
+            @second_comment.author = @author
+
+            serializer = ComplexNestedAuthorSerializer.new(@author)
+            adapter = ActiveModel::Serializer::Adapter::Json.new(serializer)
```

Comment:
personally, I feel that that what the controller does should be in separate tests. So that way if something breaks, if your tests are all separated by responsibility, it'll be easier to track down the cause.